Added support for STARTTLS.

jsf9k opened the pull request:
This PR pertains to issue #5.

The CSV output now includes ~~two~~several additional columns.  "Mail Server Ports Tested" contains a comma-delimited list of the ports that were tested.  "Mail Server Is Listening" states whether a given mail server (which comes from an MX record) and port combination is listening for TCP connections. 
"Mail Server Supports SMTP", states whether a given mail server and port combination really appears to be an SMTP server.  "Mail Server Supports STARTTLS" indicates whether the mail server and port combination supports STARTTLS.  I've run most of current-federal.csv though the code, and it appears to be working.

~~The DNS timeout serves double duty as the SMTP connection timeout
value as well.~~

jsf9k commented:
@h-m-f-t, I went ahead and created the PR.  If we uncover any issues, desired changes, etc. we can continue committing them to my branch and they will be added to the PR.

h-m-f-t commented on trustymail/trustymail.py:
```diff
@@ -63,6 +68,76 @@ def mx_scan(domain):
+            # Now try to say hello.  This will tell us if the thing
+            # that is listening is an SMTP server.
+            try:
+                smtp_connection.ehlo_or_helo_if_needed()
```
The docs for [`ehlo_or_helo_if_needed()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.ehlo_or_helo_if_needed) indicate it calls [`ehlo()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.ehlo)/[`helo()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.helo), which has `name` as a parameter. This parameter "defaults to the fully qualified domain name of the local host", which, in our and an arbitrary user's case, seems liable to be an .in-addr.arpa address or just the IP.

Per the 6th and 7th paragraphs of [section 4.1.4. of RFC 5321](https://tools.ietf.org/html/rfc5321#section-4.1.4), the current SMTP standard, this looks to be ok (if an address doesn't reverse to be an FQDN, SMTP will use the raw address). Perhaps we should run a few tests and see if hardcoding (or making configurable) an EHLO name (postmaster@.gov) makes any difference in our results?

jsf9k replied:
I will add a flag for setting the name.  By default it will work as it does now, but if you specify the name it will be used instead.

h-m-f-t commented on trustymail/trustymail.py:
```diff
@@ -63,6 +68,76 @@ def mx_scan(domain):
+
+            # We were able to connect to the server and say hello, so
+            # it must send mail.
+            domain.sends_mail.append(server_and_port + "-" + str(True))
```
We may want to play with the verbiage here; "sends mail" may be fudgy. I don't necessarily think that technical correctness should go ahead of user friendliness, but just that "sends mail" may be too general. 

Port 25 is used for mail transmission while port 587 is for _mail submission_ (as is port 465, unofficially). The standards track RFC (4409) that defines [mail submission](https://tools.ietf.org/html/rfc4409#section-3.1), and especially the [BCP on mail submission](https://tools.ietf.org/html/rfc5068#section-2), describe how mail transit and submission are different. Indeed, the [STARTTLS RFC](https://tools.ietf.org/html/rfc3207#section-4.3) calls out the submission port as being different from (and benefitting especially because) STARTTLS is operating on the submission port.

jsf9k replied:
How about "supports SMTP"?

h-m-f-t replied:
👍 yep, that works.

h-m-f-t commented on trustymail/trustymail.py:
```diff
@@ -13,14 +15,17 @@
 ]
 
+# The ports that will be checked to see if an SMTP server is listening
+SMTP_PORTS = [25, 465, 587]
```
I think we need data from a full run on port 465. I'm of a mind not to include in our ports definition unless deployment is really strong in the federal space.

jsf9k replied:
I will leave this as it is for now, and we can look at the results to determine if 465 should be removed or not.  I will also create a flag to override the default ports with a comma-separated list of ports.

h-m-f-t commented on trustymail/domain.py:
```diff
@@ -41,12 +41,34 @@ def __init__(self, domain_name):
+
+    def has_starttls(self):
+        """
+        Returns True if STARTTLS information is present and otherwise
```
I think `Has STARTTLS` should be a Boolean indicating True when (at least one port? all ports?) use STARTTLS. The raw findings should probably be called something different the `Supports STARTTLS`; the nomenclature other fields in trustymail follow for this type of output is "_x_ Results", like `STARTTLS results` .

jsf9k replied:
That was exactly my confusion.  Should "Has STARTTLS" be true if the IP listens on 25 and 587 but only supports STARTTLS on 25?  

I'm happy to split "Has STARTTLS" into a Boolean field and a results field, but I need a little guidance on what to do when one port supports STARTTLS but another (active) port does not.

jsf9k commented:
@h-m-f-t, I think the changes I just committed speak to your comments.  Let me know if you have any questions or issues, and let me know if you want to remove port 465 from the defaults.

h-m-f-t commented on trustymail/domain.py:
```diff
@@ -90,6 +116,10 @@ def generate_results(self):
                         "MX Record": self.has_mail(),
                         "Mail Servers": self.format_list(self.mail_servers),
+                        "Mail Server Ports Tested": self.format_list(list(map(lambda x:str(x), self.ports_tested))),
+                        "Mail Server Is Listening": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["is_listening"]), self.starttls_results.keys()))),
```
If we're really interested in SMTP support, what advantage does knowing `Mail Server Is Listening` bring? Does this value need to be exposed in the output?

jsf9k replied:
It doesn't have to be exposed.  I just thought it may be useful to know whether something is actually listening in some edge cases.  I'm fine removing it if you prefer that.

jsf9k replied:
@h-m-f-t, did you decide that this change is something you want?  If so I can make the change.

h-m-f-t commented on trustymail/domain.py:
```diff
@@ -90,6 +116,10 @@ def generate_results(self):
                         "Mail Servers": self.format_list(self.mail_servers),
+                        "Mail Server Ports Tested": self.format_list(list(map(lambda x:str(x), self.ports_tested))),
+                        "Mail Server Is Listening": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["is_listening"]), self.starttls_results.keys()))),
+                        "Mail Server Supports SMTP": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["supports_smtp"]), self.starttls_results.keys()))),
```
Thoughts on only includes those hostnames/ports that return as `True`?

jsf9k replied:
Do you mean something like this?
```json
"Mail Server Supports SMTP": "blah.foo.com:467,mail.blah.foo.com:467"
```
I'm fine with that.

jsf9k replied:
@h-m-f-t, did you decide that this change is something you want?  If so I can make the change.

h-m-f-t commented on trustymail/domain.py:
```diff
@@ -90,6 +116,10 @@ def generate_results(self):
+                        "Mail Server Ports Tested": self.format_list(list(map(lambda x:str(x), self.ports_tested))),
+                        "Mail Server Is Listening": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["is_listening"]), self.starttls_results.keys()))),
+                        "Mail Server Supports SMTP": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["supports_smtp"]), self.starttls_results.keys()))),
+                        "Mail Server Supports STARTTLS": self.format_list(list(map(lambda x:x + ";" + str(self.starttls_results[x]["starttls"]), self.starttls_results.keys()))),
```
Same comment as above.

jsf9k replied:
See above.

IanLee1521 commented on trustymail/cli.py:
```diff
@@ -44,16 +57,39 @@ def main():
+        smtp_localhost = None
+
+    if args["--smtp-ports"] is not None:
+        smtp_ports = list(map(lambda x:int(x), args["--smtp-ports"].split(',')))
```
Forgive my new-ness to lambda's... Is this taking the arguments to the `--smtp-ports` options and returning a list of those numbers as `int`s? If so, perhaps a list comprehension is clearer?

```python
smtp_ports = [int(port) for port in args['--smtp-ports'].split(',')]
```

jsf9k replied:
Yes, that's all it's doing.  I will change it to a list comprehension.

IanLee1521 commented on trustymail/trustymail.py:
```diff
@@ -174,15 +272,19 @@ def scan(domain_name, timeout, scan_types):
 
     # If the user didn't specify any scans then run a full scan.
-    if not (scan_types["mx"] or scan_types["spf"] or scan_types["dmarc"]):
+    if not (scan_types["mx"] or scan_types["starttls"] or scan_types["spf"] or scan_types["dmarc"]):
```
Not really related to the change, but should there also be a condition here like above for `... and domain.is_live` ?

jsf9k replied:
Good question.  @h-m-f-t?

h-m-f-t replied:
@IanLee1521, yeah, I think we do want that to add that! If the domain doesn't exist (isn't live), there's no reason to run any record- or service-specific scans on it.

IanLee1521 commented on trustymail/trustymail.py:
```diff
@@ -63,6 +70,99 @@ def mx_scan(domain):
+    for mail_server in domain.mail_servers:
+        for port in smtp_ports:
+            if port not in domain.ports_tested:
+                domain.ports_tested.append(port)
```
I wonder if it would be clearer to have `domain.ports_tested` be a `set` rather than a `list` and then you could just do: `domain.ports_tested.add(port)` and you don't need to do the conditional check of if it's in there.

IanLee1521 commented on trustymail/cli.py:
```diff
@@ -26,6 +36,9 @@
 base_domains = {}
 
+# The default ports to be checked to see if an SMTP server is listening.
+_DEFAULT_SMTP_PORTS = [25, 465, 587]
```
See my other comment, but maybe this should be a set rather than a list? 

```python
_DEFAULT_SMTP_PORTS = {25, 465, 587}
```

IanLee1521 commented on trustymail/cli.py:
```diff
@@ -44,16 +57,39 @@ def main():
+        smtp_localhost = None
+
+    if args["--smtp-ports"] is not None:
+        smtp_ports = [int(port) for port in args['--smtp-ports'].split(',')]
```
Same:

```python
smtp_ports = {int(port) for port in args['--smtp-ports'].split(',')}
```

jsf9k commented:
Great suggestions as usual, @IanLee1521!

h-m-f-t commented:
@jsf9k @IanLee1521 excellent additions! I really appreciate your efforts.

I did a full run on [current-federal](https://github.com/GSA/data/blob/master/dotgov-domains/current-federal.csv) with all the defaults, and I think the thing that's most difficult from a usability standpoint is the representation of the "Mail Server *" fields. It's hard to get an at-a-glance view of whether a `Mail Server is Listening` or `Supports STARTTLS` or `Supports SMTP`. I totally get this nuance is inherent in us checking multiple ports, but wonder if maybe we should "arbitrarily define" what those fields mean, making them a clear Boolean, and defining them away in the README as "`Supports...`" means that the mail server is responsive on a given port", with the actual `Results` in a separate field. This not only makes the tool more user-friendly, it would make our report-generation code need to parse less.

konklone commented:
Since any run of `pshtt` will have the same ports tested for each row, and the hostnames won't differ, how about instead of a `Mail Server Ports Tested` field and a `Mail Server Is Listening` field, have one field with a value of whatever ports the hostname is listening on? So the value could just be `25,465` or `465`.

jsf9k commented:
@h-m-f-t, I liked your previous suggestion of only including those entries that end up being true.  So `Mail Server Supports SMTP` would look like `blah.blah.com:25, blah.blah.com:457`, for example.  I feel like reducing that to just a single Boolean value throws away a lot of information that in many cases would be useful to the user.

That said, I'll agree to whatever the group agrees to.

jsf9k commented:
@konklone, a single domain can have multiple MX records.  `gmail.com`, for instance, returns five of them.  Each MX record is a potential mail server hostname.  I think we have to include the hostname + port combinations since there can be more than one mail server hostname per row.

jsf9k commented:
I went ahead and implemented @h-m-f-t's original suggestion.  I also fixed a minor bug where the in-memory cache was being built (but not used) even if the user did not want to use the cache.

jsf9k commented:
After discussion with @h-m-f-t, I added an extra boolean field to indicate the domain's overall compliance with STARTTLS.

h-m-f-t commented on README.md:
```diff
@@ -69,8 +79,18 @@ The following values are returned in `results.csv`:
 * `Mail Servers` - The list of hosts found in the MX record.
+* `Mail Server Ports Tested` - A list of the ports tested for SMTP and
+  STARTTLS support.
+* `Mail Server Is Listening` - If the mail servers are actually
```
What does `Mail Server Is Listening` provide that `Mail Server Supports SMTP` doesn't?

jsf9k replied:
In order to try to speak SMTP to a host on a given port I have to first connect via TCP.  If that succeeds then I know _something_ is listening at that host on that port, whether it's smtp, telnet, ssh, or something else.  Since I get that information for free I was including it, but I could just as easily not record it.

I can imagine cases where the information is useful, but it may be more confusing to leave it in. 
 In any event, the same information is duplicated (and expanded upon) in the CYHY report.

h-m-f-t replied:
We change our mind, but I agree dropping it makes sense for now.

h-m-f-t commented on README.md:
```diff
@@ -69,8 +79,18 @@ The following values are returned in `results.csv`:
+  STARTTLS support.
+* `Mail Server Is Listening` - If the mail servers are actually
+  listening on the tested ports.
+* `Mail Server Supports SMTP` - If the mail servers support SMTP.
```
I propose that the field `Mail Server Supports SMTP` become Boolean (true if at least one domain on the MX record "speaks" SMTP), with the current `Mail Server Supports SMTP` becoming `Mail Server Supports SMTP Results`.

h-m-f-t replied:
Maybe it makes sense to rename both fields from `Mail Server Supports...` to `Domain Supports...`, since it's semantically odd for a `*Mail Server* Supports SMTP` to be false (for a "mail server" to not offer SMTP).

jsf9k replied:
:+1:

h-m-f-t replied:
@jsf9k last thing: let's put the `...Supports` Boolean *before* the `...Supports Results` field, which is helpful from a human-using-the-tool-output use case.

h-m-f-t commented on README.md:
```diff
@@ -69,8 +79,18 @@ The following values are returned in `results.csv`:
+* `Mail Server Is Listening` - If the mail servers are actually
+  listening on the tested ports.
+* `Mail Server Supports SMTP` - If the mail servers support SMTP.
+* `Mail Server Supports STARTTLS` - If the mail servers support
```
Same comments as above. `Mail Server Supports STARTTLS` becomes Boolean (true if only all SMTP-offering endpoints enable STARTTLS`, and `Mail Server Supports STARTTLS Results` listing the results.

Also, same comment that maybe `Domain Supports...` is better language here, too.

jsf9k replied:
:+1:

jsf9k commented:
@h-m-f-t, if you're happy with the changes I made go ahead and merge.

jsf9k commented:
Yahoo!  :metal:

IanLee1521 commented on trustymail/domain.py:
```diff
@@ -83,30 +109,42 @@ def get_dmarc_policy(self):
 
 
     def generate_results(self):
+        mail_servers_that_are_listening = [x for x in self.starttls_results.keys() if self.starttls_results[x]["is_listening"]]
```
FYI -- Apparently `mail_servers_that_are_listening` isn't actually being used.

jsf9k replied:
Good catch, @IanLee1521!  We removed that field from the output yesterday.  I'll create a new pull request for this.

IanLee1521 replied:
Found it while flake8-ifying the code. Will be opening that PR in a bit.

IanLee1521 replied:
Done in https://github.com/dhs-ncats/trustymail/pull/20
